File: backend/app/schemas/Portfolio.py
from pydantic import BaseModel, field_validator
from datetime import datetime
from decimal import Decimal


class PortfolioHoldingResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    ticker: str
    target_pct: Decimal


class PortfolioHoldingHistoryResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    ticker: str
    change_type: str
    target_pct: Decimal
    old_target_pct: Decimal | None
    changed_at: datetime


class PortfolioResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    name: str
    description: str | None
    icon_url: str | None
    is_active: bool
    total_return_pct: Decimal | None
    return_1w: Decimal | None
    return_1m: Decimal | None
    return_3m: Decimal | None
    return_6m: Decimal | None
    return_1y: Decimal | None
    return_3y: Decimal | None
    holdings: list[PortfolioHoldingResponse] = []
    created_at: datetime
    user_count: int = 0


class AdminPortfolioResponse(PortfolioResponse):
    holding_history: list[PortfolioHoldingHistoryResponse] = []
    total_invested: str = "0.00"


class MarketplaceResponse(BaseModel):
    portfolios: list[PortfolioResponse]
    total: int
    page: int
    page_size: int


class CreatePortfolioPayload(BaseModel):
    name: str
    description: str | None = None
    icon_url: str | None = None


class UpdatePortfolioPayload(BaseModel):
    name: str | None = None
    description: str | None = None
    icon_url: str | None = None


class UpdateHoldingsPayload(BaseModel):
    holdings: list[dict]   # [{"ticker": "AAPL", "target_pct": "45.00"}, ...]


class JoinPortfolioPayload(BaseModel):
    portfolio_id: int
    brokerage_connection_id: int | None = None
    investment_amount: Decimal | None = None

    @field_validator("investment_amount")
    @classmethod
    def investment_amount_minimum(cls, v: Decimal | None) -> Decimal | None:
        if v is not None and v < Decimal("100"):
            raise ValueError("Minimum investment amount is $100.00")
        return v


class PaginatedPortfoliosResponse(BaseModel):
    portfolios: list[PortfolioResponse]
    total: int
    page: int
    page_size: int


class PaginatedAdminPortfoliosResponse(BaseModel):
    portfolios: list[AdminPortfolioResponse]
    total: int
    page: int
    page_size: int

File: frontend/models/Portfolio.ts
export interface PortfolioHolding {
  id: number;
  ticker: string;
  target_pct: string;
}

export interface PortfolioHoldingHistory {
  id: number;
  ticker: string;
  change_type: 'added' | 'removed' | 'changed';
  target_pct: string;
  old_target_pct: string | null;
  changed_at: string;
}

export interface Portfolio {
  id: number;
  name: string;
  description: string | null;
  icon_url: string | null;
  is_active: boolean;
  total_return_pct: string | null;
  return_1w: string | null;
  return_1m: string | null;
  return_3m: string | null;
  return_6m: string | null;
  return_1y: string | null;
  return_3y: string | null;
  holdings: PortfolioHolding[];
  created_at: string;
  user_count?: number;
}

export interface AdminPortfolio extends Portfolio {
  holding_history: PortfolioHoldingHistory[];
  user_count: number;
  total_invested: string;
}

export interface CreatePortfolioPayload {
  name: string;
  description?: string;
  icon_url?: string;
}

export interface UpdatePortfolioPayload {
  name?: string;
  description?: string;
  icon_url?: string;
}

export interface UpdateHoldingsPayload {
  holdings: { ticker: string; target_pct: string }[];
}

export interface JoinPortfolioPayload {
  portfolio_id: number;
  brokerage_connection_id?: number;
  investment_amount?: number;
}

export interface PaginatedPortfolios {
  portfolios: Portfolio[];
  total: number;
  page: number;
  page_size: number;
}

export interface PaginatedAdminPortfolios {
  portfolios: AdminPortfolio[];
  total: number;
  page: number;
  page_size: number;
}
